### ISSUE TITLE
Streaming

### ISSUE THREAD

**Issue by flexgames** [jonasman/TeslaSwift#6]
This isn't an issue per say, but can TeslaSwift support the TESLA HTTP Long Polling "STREAMING" API to get continuous telemetry from the Tesla Model S? I would like to be able to do this without hammering the DriveState class (as I think that generates a login call to Tesla's servers each time). Thanks.

**Comment by jonasman** [comment]
Hi
Drivestate does just ine request after login is done.
What streaming are you talking? The same as is used for summon?

**Comment by flexgames** [comment]
Hi João

On:
https://github.com/hjespers/teslams <https://github.com/hjespers/teslams>

Check out the streaming.js section in that link.

It provides a continuous stream of location data without hammering the server with json requests (I believe). I have the .js files working on my local Mac but would like to write my own Mac app with Swift and your library seems to be a perfect fit, with this one possible exception.

Thanks,

Joe

**Comment by jonasman** [comment]
thanks for the heads up. Seems to be easy to add to the library as it is just a special request.

**Comment by jonasman** [comment]
there is another example here https://github.com/timdorr/model-s-api/blob/master/lib/tesla_api/stream.rb

**Comment by jakeholland** [comment]
@jonasman - I currently have a project where I would like to use the streaming API and have started the work to add it in. Just wanted to check that I am not duplicating efforts. 😄

**Comment by flexgames** [comment]
@jakeholland Let me know if you get anywhere with this. I'd be interested in using it if you proceed. I am doing a bit of a workaround right now to get the functionality that you can read about in the bottom of this thread:
[model-s-api](https://github.com/timdorr/model-s-api/issues/16#issuecomment-284128531)

**Comment by jonasman** [comment]
This is in the TODO of this lib. But not ETA yet. I need free time to make it. :)

**Comment by jakeholland** [comment]
I will get to work on it, have a lot of experience with web sockets. Problem I have is that I don't have a model X/S. I am waiting on a Model 3. So some stuff might be a bit of guess and check if anybody wants to work with me on testing that would be awesome!

**Comment by flexgames** [comment]
I'd be more than glad to test it with mine!

**Comment by jonasman** [comment]
If you could do a PR i could test it on my car

**Comment by jakeholland** [comment]
@flexgames - I wrote some simple code to open a socket connection with the streaming API earlier today. However, I was just looking back through that thread and it looks pretty similar to what you had. Just looking briefly at your code, I'm not sure my code would be any different, but I'll create a PR so you guys can check if out if you want.

**Comment by jakeholland** [comment]
@flexgames - FYI PR created here: https://github.com/jonasman/TeslaSwift/pull/9

**Comment by flexgames** [comment]
Awesome thanks! I'm flying out tonight but will test when I return later this week. Interesting re your comment about ws vs wss. I have to go back and check my branch and see if that fixes it.

**Comment by jonasman** [comment]
So i picked on this today and have a starting point code from @jakeholland 's PR .

I fixed a couple of bugs there to make the stream to connect but it seems like i cannot due to TESLA sending a 200 OK instead of a 101 code. not sure what is the problem :)

I tried also StarScream websocket just to make sure the problem is not the websocket lib.

**Comment by jonasman** [comment]
connecting to /connect/ endpoint works. (it is the autopark/summon endpoint)

**Comment by jonasman** [comment]
I created a new branch based on #9 called streamingAPI. This branch contains my progress on connecting to the streaming API. It still doesn't work because of the errors mentioned above.

**Comment by jonasman** [comment]
Quick update: It seems like Tesla is not using websockets for its streaming API but instead HTTP SSE.

The streamingAPI branch includes all code ready to test this connection. But there's no data parsing.
My car doesn't stream any thing. maybe i need to drive it to get any streaming data?

This is different than summon APIs. And those will be covered in other issues.

**Comment by flexgames** [comment]
Mine streams, however it streams more slowly when parked. I can test it.

**Comment by jonasman** [comment]
Are you streaming using websockets or http sse?

**Comment by flexgames** [comment]
Sorry I should be more clear - mine streams with the Node.js code, and I sample it from my Swift app at the moment.

**Comment by jonasman** [comment]
Is your node code webstockets or http sse?

**Comment by flexgames** [comment]
I am not sure. Here is the specific javascript that is being used:

[hjespers/teslams](https://github.com/hjespers/teslams/blob/master/examples/streaming.js)

**Comment by jonasman** [comment]
went for a drive, same thing. No events :(

**Comment by jakeholland** [comment]
Just read through all your comments, sorry I can't be of much help at the moment. ☹️ I looked through a few apis and also found this some information on this page and a curl command for the streaming API https://teslamotorsclub.com/tmc/threads/model-s-rest-api.13410/page-5

**Comment by jonasman** [comment]
Good news. I managed to make the streaming api to work
You can review the code here: https://github.com/jonasman/TeslaSwift/pull/10
It is not pretty but works
